[Fix #599] Add Rails/DuplicateAssociation cop

natematykiewicz opened the pull request:
Fixes #599

Add Rails/DuplicateAssociation cop

-----------------

Before submitting the PR make sure the following are checked:

* [X] The PR relates to *only* one subject with a clear title and description in grammatically correct, complete sentences.
* [X] Wrote [good commit messages][1].
* [X] Commit message starts with `[Fix #issue-number]` (if the related issue exists).
* [X] Feature branch is up-to-date with `master` (if not - rebase it).
* [X] Squashed related commits together.
* [X] Added tests.
* [X] Ran `bundle exec rake default`. It executes all tests and runs RuboCop on its own code.
* [X] Added an entry (file) to the [changelog folder](https://github.com/rubocop/rubocop-rails/blob/master/changelog/) named `{change_type}_{change_description}.md` if the new code introduces user-observable changes. See [changelog entry format](https://github.com/rubocop/rubocop/blob/master/CONTRIBUTING.md#changelog-entry-format) for details.
* [X] The PR relates to *only* one subject with a clear title
  and description in grammatically correct, complete sentences.
* [X] If this is a new cop, consider making a corresponding update to the [Rails Style Guide](https://github.com/rubocop/rails-style-guide).

[1]: https://chris.beams.io/posts/git-commit/

natematykiewicz commented:
Hey @pirj! Continuing our conversation in #599!

This is my first Rubocop PR. I had some errors with the matcher you provided. From what I can tell, constants aren't supported in the matcher. For example, the [inverse of](https://github.com/rubocop/rubocop-rails/blob/master/lib/rubocop/cop/rails/inverse_of.rb#L134-L138) cop defined the `RESTRICT_ON_SEND` constant with a few association names, but then retypes the association names in the node matcher.

I copied the `corrector` by searching for `corrector.remove` and following what seemed relevant.

I couldn't figure out how to use the duplication helper for this.

The `class_send_nodes` came from [after_commit_override](https://github.com/rubocop/rubocop-rails/blob/master/lib/rubocop/cop/rails/after_commit_override.rb#L66-L76).

pirj reviewed (approved):
Looks flawless, thank you!

After giving it another thought, maybe it doesn't make sense to consider the association method in uniqueness, so that the following is detected:
```ruby
has_and_belongs_to_many :bars
has_many :bars
```
ca be detected.

There can be two-way associations, but they can't be described using the same key anyway:
```ruby
has_one :user
belongs_to :user
```

pirj commented:
CI failure is flaky, please disregard. We should add `gem update bundler` to our buildfile. See https://github.com/rubocop/rubocop-rails/issues/597

natematykiewicz commented:
> After giving it another thought, maybe it doesn't make sense to consider the association method in uniqueness, so that the following is detected:
> 
> ```ruby
> 
> has_and_belongs_to_many :bars
> 
> has_many :bars
> 
> ```
> 
> can be detected.
> 

Great point! I'll get that going. Shouldn't be too hard.

natematykiewicz commented:
@pirj I just pushed a few commits. It now makes sure that only 1 association (regardless of association type) is defined with a single name. I also added the offending association name to the error message, just to make the problem more clear.

pirj commented on lib/rubocop/cop/rails/duplicate_association.rb:
```diff
@@ -0,0 +1,67 @@
+
+        def offenses(class_node)
+          class_send_nodes(class_node).select { |node| association(node) }
+                                      .group_by { |assoc| association_name(assoc) }
```
`association` node matcher captures (`$`) type and name of associations.
If you call `association(assoc)`, it will return something like `[:belongs_too, :foo]`.
You may simplify code a bit by:
 - removing capture (`$`) from `${:belongs_to`, as we don't use association type anymore
 - using `group_by` by the same `association` and remove `association_name` method

pirj replied:
![image](https://user-images.githubusercontent.com/6916/147555457-0eba77d4-063f-4a17-ab41-e6f6fb40aa2e.png)

natematykiewicz replied:
Makes sense. Thanks!

pirj commented on lib/rubocop/cop/rails/duplicate_association.rb:
```diff
@@ -0,0 +1,67 @@
+          if class_def.send_type?
+            [class_def]
+          else
+            class_def.each_child_node(:send).to_a
```
Is `to_a` necessary? I think `each_child_node` returns an iterator.

natematykiewicz replied:
It's not. Should I remove it from [after_commit_override](https://github.com/rubocop/rubocop-rails/blob/master/lib/rubocop/cop/rails/after_commit_override.rb#L74) too then? That's where I copied it from.

Or, should I move this shared code somewhere else to DRY it? It seems really helpful for any time you want to capture a bunch of class method calls in `on_class`.

pirj replied:
Fair enough. 👍 
Yes, `to_a` looks redundant there, too.
Please have no hesitation to extract `class_send_nodes` into a mixin if you're up for it.

natematykiewicz replied:
I could do that! Would that be a follow up PR or part of this one? Or perhaps I do it in the duplicate scopes PR?

pirj replied:
The duplicate scopes PR seems to be a reasonable choice 👍

natematykiewicz replied:
Sounds good to me! I'll do that after this one gets merged.

pirj commented on lib/rubocop/cop/rails/duplicate_association.rb:
```diff
@@ -0,0 +1,67 @@
+          offenses(class_node).each do |name, nodes|
+            length = nodes.length
+
+            nodes.each.with_index do |node, i|
```
Maybe `nodes[0..-2]` and remove `length`/`with_index`/`i`/`if`?

pirj replied:
Please disregard this. I've missed the fact that we raise offences for all, but remove all except the last one on autocorrection.

Alternatively,
![image](https://user-images.githubusercontent.com/6916/147556086-46ca6868-070b-4ec1-bee8-8f076151b0b9.png)

Leaving this up to you if you'd like to keep this as is.

natematykiewicz replied:
Nice! That line's too long. Moving it to a guard clause.

pirj reviewed (approved):
Fantastic, thanks!

natematykiewicz commented:
@pirj I've addressed all of the comments. Unless you have any more feedback (which I'll gladly hear), I think this question is the only remaining thing: https://github.com/rubocop/rubocop-rails/pull/602#discussion_r776158099

natematykiewicz commented:
I made a duplicate scope cop since someone asked for it and it was loosely the same thing as this. There's a lot of similarity between these 2 PRs though. https://github.com/rubocop/rubocop-rails/pull/607/files#diff-44c86a2e4d44563d60e60596b9063457f2568af543209f8d8c6489581d33cb1a

I keep looking at this [Duplication mixin](https://github.com/rubocop/rubocop/blob/1e55b1aa5e4c5eaeccad5d61f08b7930ed6bc341/lib/rubocop/cop/mixin/duplication.rb) you mentioned. It doesn't seem to help because it simply takes a collection and returns elements that are duplicated. I need to take a node list and determine which are duplicated based on their name (or in the scope PR, duplicate based on the proc arg).

pirj commented:
Fantastic! Thanks a lot for your dedication.
I have nothing to add, the PR looks perfect to me. Thank you!

natematykiewicz commented:
Awesome! I just converted #607 to a draft to deal with the `class_send_nodes` mixin over there.

natematykiewicz commented:
@pirj What's left to get this merged?

pirj commented:
@koic do you have some feedback on this one?

koic commented on lib/rubocop/cop/rails/duplicate_association.rb:
```diff
@@ -0,0 +1,63 @@
+      #
+      # @example
+      #
+      #   # bad
+      #   belongs_to :foo
+      #   belongs_to :bar
+      #   has_one :foo
+      #
+      #   # good
+      #   belongs_to :foo
+      #   has_one :bar
```
This example looks `bad` and `good` cases are incompatible. It looks like it should be:

```suggestion
      #   # bad
      #   belongs_to :foo
      #   belongs_to :bar
      #   has_one :foo
      #
      #   # good
      #   belongs_to :bar
      #   has_one :foo
```

How about?

natematykiewicz replied:
🤦🏻‍♂️ Fixed. Thanks

koic commented on lib/rubocop/cop/rails/duplicate_association.rb:
```diff
@@ -0,0 +1,63 @@
+module RuboCop
+  module Cop
+    module Rails
+      # This cop looks for associations that have been defined multiple times in the same file.
```
How about adding to the documentation that auto-correction uses the last defined association in duplicated same associations for compatibility?

natematykiewicz replied:
Done! Let me know what you think.

koic commented on lib/rubocop/cop/rails/duplicate_association.rb:
```diff
@@ -0,0 +1,63 @@
+        include RangeHelp
+        extend AutoCorrector
+
+        MSG = "Association `%<name>s` is defined multiple times. Don't repeat associations in a model."
```
Can you tweak the message?

```suggestion
        MSG = "Association `%<name>s` is defined multiple times. Don't repeat associations."
```

natematykiewicz replied:
Done!

natematykiewicz commented:
@koic I made those changes you suggested, and also rebased to master.

koic commented on lib/rubocop/cop/rails/duplicate_association.rb:
```diff
@@ -0,0 +1,67 @@
+    module Rails
+      # This cop looks for associations that have been defined multiple times in the same file.
+      #
+      # When an association is defined multiple times on a model, ActiveRecord overrides the
```
Nice!
This is a nitpick. Maybe "Active Record" seems to be suitable this one.

> The proper names of Rails components have a space in between the words, like "Active Support". `ActiveRecord` is a Ruby module, whereas Active Record is an ORM. 

https://guides.rubyonrails.org/api_documentation_guidelines.html#wording

```suggestion
      # When an association is defined multiple times on a model, Active Record overrides the
```

natematykiewicz replied:
Done!

koic commented on changelog/new_add_railsduplicateassociation_cop.md:
```diff
@@ -0,0 +1 @@
+* [#599](https://github.com/rubocop/rubocop-rails/issues/599): Add Rails/DuplicateAssociation cop. ([@natematykiewicz][])
```
Can you tweak to the following?

```suggestion
* [#599](https://github.com/rubocop/rubocop-rails/issues/599): Add new `Rails/DuplicateAssociation` cop. ([@natematykiewicz][])
```

natematykiewicz replied:
Done!

koic commented:
I've left small feedback comments, but overall this looks good to me. Can you squash your commits into one?

natematykiewicz commented:
@koic I've made those tweaks and squashed to 1 commit

koic commented:
This looks good to me 🌟 I will merge this cop after the next bug fix release.

koic commented:
@natematykiewicz Can you rebase with the latest master branch?

natematykiewicz commented:
Done @koic!

koic commented:
Thanks!

pirj commented:
@natematykiewicz Thanks for your contribution and persistence!
